src/lib/logs/log_integration_service.py

```diff
@@ -0,0 +1,117 @@
+from lib.logs.dynatrace_client import DynatraceClient
+from lib.logs.gcp_client import GCPClient
+from lib.logs.log_self_monitoring import put_sfm_into_queue
+from lib.logs.logs_processor import (
+    prepare_context_and_process_message,
+    prepare_batches,
+    LogBatch,
+)
+from lib.utilities import chunks
+from lib.logs.log_forwarder_variables import (
+    REQUEST_BODY_MAX_SIZE,
+    NUMBER_OF_CONCURRENT_MESSAGE_PULL_COROUTINES,
+    NUMBER_OF_CONCURRENT_PUSH_COROUTINES,
+)
+
+
+class LogIntegrationService:
+    def __init__(self, sfm_queue: asyncio.Queue):
+        self.gcp_client = None
+        self.dynatrace_client = DynatraceClient()
+        self.sfm_queue = sfm_queue
+        self.log_push_semaphore = asyncio.Semaphore(NUMBER_OF_CONCURRENT_PUSH_COROUTINES)
+
+    async def update_gcp_client(self, logging_context: LoggingContext):
+        async with init_gcp_client_session() as gcp_session:
+            token = await create_token(logging_context, gcp_session)
+            if token is None:
+                raise Exception("Cannot start pub/sub pulling - Failed to fetch token")
+            self.gcp_client = GCPClient(token)
+
+    async def keep_gcp_token_updated(self, logging_context: LoggingContext):
+        while True:
+            await asyncio.sleep(50 * 60)
+            task = self.update_gcp_client(logging_context)
+            try:
+                await asyncio.wait_for(task, 5 * 60)
+            except asyncio.exceptions.TimeoutError as e:
+                raise Exception("Failed to fetch Google API token")
+
+    async def perform_pull(
+        self, logging_context: LoggingContext
+    ) -> Tuple[List[LogBatch], List[str]]:
+        async with init_gcp_client_session() as gcp_session:
+            context = LogsProcessingContext(None, None, self.sfm_queue)
+            context.self_monitoring.pulling_time_start = time.perf_counter()
+
+            tasks_to_pull_messages = [
+                self.gcp_client.pull_messages(logging_context, gcp_session)
+                for _ in range(NUMBER_OF_CONCURRENT_MESSAGE_PULL_COROUTINES)
+            ]
+            responses = await asyncio.gather(*tasks_to_pull_messages, return_exceptions=True)
+
+            context.self_monitoring.calculate_pulling_time()
+            processed_messages = []
+            ack_ids_to_send = []
+
+            context.self_monitoring.processing_time_start = time.perf_counter()
+            for response in responses:
+                if not isinstance(response, Exception):
+                    for received_message in response.get("receivedMessages", []):  # type: ignore
+                        message_job = prepare_context_and_process_message(
+                            self.sfm_queue, received_message
+                        )
+                        if not message_job or message_job.bytes_size > REQUEST_BODY_MAX_SIZE - 2:
+                            ack_ids_to_send.append(received_message.get("ackId"))
```

Comment:
I think this for loop should be less complex. Could you consider for example creating an additional function to separate the logic?